extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/glyptodon/guacamole/auth/jdbc/socket/AbstractGuacamoleSocketService.java

```diff
@@ -217,25 +301,20 @@ public void close() throws GuacamoleException {
                     // Attempt to close connection
                     super.close();
                     
-                    // Release connection upon close
-                    activeConnections.remove(identifier, activeConnection);
-                    activeConnectionGroups.remove(parentIdentifier, activeConnection);
-                    release(user, connection);
-
-                    UserModel userModel = user.getUser().getModel();
-                    ConnectionRecordModel recordModel = new ConnectionRecordModel();
-
-                    // Copy user information and timestamps into new record
-                    recordModel.setUserID(userModel.getObjectID());
-                    recordModel.setUsername(userModel.getIdentifier());
-                    recordModel.setConnectionIdentifier(identifier);
-                    recordModel.setStartDate(activeConnection.getStartDate());
-                    recordModel.setEndDate(new Date());
-
-                    // Insert connection record
-                    connectionRecordMapper.insert(recordModel);
+                    // Release connection upon close, if not already released
+                    if (released.compareAndSet(false, true)) {
+
+                        // Release connection
+                        activeConnections.remove(identifier, activeConnection);
+                        activeConnectionGroups.remove(parentIdentifier, activeConnection);
+                        release(user, connection);
+
+                        // Save record to database
+                        saveConnectionRecord(identifier, activeConnection);
+
+                    }
                     
-                }
+                } // end close()
```

Comment:
It's hard to see from the diff, but I added this due to the number of nested braces. There are four levels, which is hard to look at.

I added this roughly in the middle to make it easier to see what ends what.

The full context looks like:

```java

            // Return newly-reserved connection
            return new ConfiguredGuacamoleSocket(
                getUnconfiguredGuacamoleSocket(),
                getGuacamoleConfiguration(user, connection),
                info
            ) {

                @Override
                public void close() throws GuacamoleException {

                    // Attempt to close connection
                    super.close();
                    
                    // Release connection upon close, if not already released
                    if (released.compareAndSet(false, true)) {

                        // Release connection
                        activeConnections.remove(identifier, activeConnection);
                        activeConnectionGroups.remove(parentIdentifier, activeConnection);
                        release(user, connection);

                        // Save record to database
                        saveConnectionRecord(identifier, activeConnection);

                    }
                    
                } // end close()
                
            };

        }

        // Release connection in case of error
        catch (GuacamoleException e) {
```